examples/src/main/java/com/squareup/moshi/recipes/Unwrap.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi.recipes;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonQualifier;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import com.squareup.moshi.recipes.Unwrap.EnvelopJsonAdapter.Enveloped;
+import com.squareup.moshi.recipes.models.Card;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Set;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+final class Unwrap {
+  @Enveloped private static Object unused;
+
+  public static void main(String[] args) throws Exception {
+    String json = ""
+        + "{\"data\":"
+        + "  {\n"
+        + "    \"rank\": \"4\",\n"
+        + "    \"suit\": \"CLUBS\"\n"
+        + "  }"
+        + "}";
+    Annotation annotation = Unwrap.class.getDeclaredField("unused").getAnnotations()[0];
+    Moshi moshi = new Moshi.Builder().add(EnvelopJsonAdapter.FACTORY).build();
+    JsonAdapter<Card> adapter =
+        moshi.adapter(Card.class, Collections.<Annotation>singleton(annotation));
+    Card out = adapter.fromJson(json);
+    System.out.println(out);
+  }
+
+  public static final class EnvelopJsonAdapter extends JsonAdapter<Object> {
+    public static final JsonAdapter.Factory FACTORY = new Factory() {
+      @Override
+      public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
+        if (!hasJsonQualifier(annotations, Enveloped.class)) {
+          return null;
+        }
+        Type envelope =
+            Types.newParameterizedTypeWithOwner(EnvelopJsonAdapter.class, Envelope.class,
+                Types.getRawType(type));
+        JsonAdapter<Envelope<?>> delegate =
+            moshi.nextAdapter(this, envelope, Collections.<Annotation>emptySet());
+        return new EnvelopJsonAdapter(delegate);
+      }
+    };
+
+    @Retention(RUNTIME) @JsonQualifier @interface Enveloped {
+    }
+
+    private static final class Envelope<T> {
```

Comment:
Yes, I considered this, and it's what I do at work, but I thought it might get weird for this example. Cool, I'll leave the actual envelope class here, then.